allennlp/models/semantic_parsing/ratecalculus/ratecalculus_semantic_parser.py

```diff
@@ -0,0 +1,682 @@
+        self._num_start_types = 1  # TODO(mattg): get this in a more principled way somehow?
+        self._embedding_dim = question_embedder.get_output_dim()
+        self._type_params = torch.nn.Linear(self._num_entity_types, self._embedding_dim)
+        self._neighbor_params = torch.nn.Linear(self._embedding_dim, self._embedding_dim)
+
+        if num_linking_features > 0:
+            self._linking_params = torch.nn.Linear(num_linking_features, 1)
+        else:
+            self._linking_params = None
+
+        if self._use_neighbor_similarity_for_linking:
+            self._question_entity_params = torch.nn.Linear(1, 1)
+            self._question_neighbor_params = torch.nn.Linear(1, 1)
+        else:
+            self._question_entity_params = None
+            self._question_neighbor_params = None
+
+    def _get_initial_state_and_scores(self,
+                                      question: Dict[str, torch.LongTensor],
+                                      question_knowledge_graph: Dict[str, torch.LongTensor],
+                                      world: List[RateCalculusWorld],
+                                      actions: List[List[ProductionRuleArray]],
+                                      example_lisp_string: List[str] = None,
+                                      add_world_to_initial_state: bool = False,
+                                      checklist_states: List[ChecklistState] = None) -> Dict:
+        """
+        Does initial preparation and creates an intiial state for both the semantic parsers. Note
+        that the checklist state is optional, and the ``RateCalculusMmlParser`` is not expected to
+        pass it.
+        """
+        kg_text = question_knowledge_graph['text']
+        # (batch_size, question_length, embedding_dim)
+        embedded_question = self._question_embedder(question)
+        question_mask = util.get_text_field_mask(question).float()
+        # (batch_size, num_entities, num_entity_tokens, embedding_dim)
+        embedded_kg = self._question_embedder(kg_text, num_wrapping_dims=1)
+        kg_mask = util.get_text_field_mask(kg_text, num_wrapping_dims=1).float()
+
+        batch_size, num_entities, num_entity_tokens, _ = embedded_kg.size()
+        num_question_tokens = embedded_question.size(1)
+
+        # (batch_size, num_entities, embedding_dim)
+        encoded_kg = self._entity_encoder(embedded_kg, kg_mask)
+        # (batch_size, num_entities, num_neighbors)
+        neighbor_indices = self._get_neighbor_indices(world, num_entities, encoded_kg)
+
+        # Neighbor_indices is padded with -1 since 0 is a potential neighbor index.
+        # Thus, the absolute value needs to be taken in the index_select, and 1 needs to
+        # be added for the mask since that method expects 0 for padding.
+        # (batch_size, num_entities, num_neighbors, embedding_dim)
+        embedded_neighbors = util.batched_index_select(encoded_kg, torch.abs(neighbor_indices))
```

Comment:
I think this is assuming that `num_entities` is at least 2. Because if it is 1, and `encoded_kg`'s first non-batch dimension is of length 1, the padding would cause `batched_index_select` to throw an out of bounds error.
How about something like the following:
```
masked_neighbor_indices = neighbor_indices * (neighbor_indices =! -1).long
embedded_neighbors = util.batched_index_select(encoded_kg, masked_neighbor_indices)
```